train/train_process.py

```diff
@@ -173,103 +176,144 @@ def _self_play(self, model_path):
-        train_dataset = train_dataset.batch(self.cfg["training"]["batch_size"], drop_remainder=True)
-        train_dataset = train_dataset.map(parser.get_after_batch_reshape_func())
-        train_dataset = train_dataset.prefetch(4)
 
         lr = self.lr_scheduler.get_lr(iter_num)
-        logging.debug("Training with learning rate %f", lr)
-        keras.backend.set_value(model.optimizer.learning_rate, lr)
-
-        logging.info("Fitting new model...")
-        training_start_time = time.time()
-        history = model.fit(train_dataset, epochs=1, verbose=0).history
-        self.metrics.update(
-            {
-                "value_loss": history["value_head_loss"][0],
-                "policy_loss": history["policy_head_loss"][0],
-                "value_accuracy": history["value_head_value_head_accuracy"][0],
-                "policy_accuracy": history["policy_head_policy_head_accuracy"][0],
-                "training_duration": time.time() - training_start_time,
-            }
-        )
-
-        logging.info("Value loss {:.4f}".format(self.metrics["value_loss"]))
-        logging.info("Policy loss {:.4f}".format(self.metrics["policy_loss"]))
-        logging.info("Value accuracy {:.4f}".format(self.metrics["value_accuracy"]))
-        logging.info("Policy accuracy {:.4f}".format(self.metrics["policy_accuracy"]))
-
-        return self._save_model(model)
-
-    def _compare_models(self, best_model, latest_model):
+        logging.debug("Training models with learning rate %f", lr)
+
+        values_loss = [None] * len(models)
+        policy_loss = [None] * len(models)
+        value_accuracy = [None] * len(models)
+        policy_accuracy = [None] * len(models)
+        training_duration = [None] * len(models)
+        trained_models = [None] * len(models)
+
+        def train_models(model_list):
+            for (model_idx, (model, _model_path)) in model_list:
+                parser = DataParser(self.game, train_data_dir, self.cfg)
+                train_dataset = parser.create_train_dataset()
+
+                keras.backend.set_value(model.optimizer.learning_rate, lr)
+
+                training_start_time = time.time()
+                history = model.fit(train_dataset, epochs=1, verbose=0).history
+                train_dur = time.time() - training_start_time
+
+                values_loss[model_idx] = history["value_head_loss"][0]
+                policy_loss[model_idx] = history["policy_head_loss"][0]
+                value_accuracy[model_idx] = history["value_head_value_head_accuracy"][0]
+                policy_accuracy[model_idx] = history["policy_head_policy_head_accuracy"][0]
+                training_duration[model_idx] = train_dur
+
+                trained_models[model_idx] = (model, self._save_model(model))
+
+        # Divide the training into jobs
+        models = list(enumerate(models))
+        workers_num = min(self.cfg["training"].get("threads", 1), len(models))
+        jobs = []

```

Comment:
neat multithreading code. what do you think about separating it out? `_multithread(train_models, models, workers_num)`
including the `join`

Reply:
Also I think something like this should work:
<img width="720" alt="image" src="https://user-images.githubusercontent.com/4618146/206856700-abd24af8-181a-44bd-91b1-c742f4ba18b4.png">
(taken from https://superfastpython.com/threadpool-python/)